app/code/community/Payone/Core/Model/Service/TransactionStatus/Execute.php

```diff
@@ -32,160 +32,234 @@
+     * @return bool
      */
-    protected function getProcessedIds()
+    protected function hasExecutionTime()
     {
-        return array_map(function ($processed) {
-            return (int) $processed->getId();
-        }, $this->processed);
+        // Check if we having remaining execution time,
+        // use a safe gap of 2 seconds
+        return is_int($this->startTime) && is_int($this->maxExecutionTime)
+            ? (time() - $this->startTime - 2) < $this->maxExecutionTime
+            : false;
     }
 
     /**
-     * @return int[] The IDs of all failed transaction status.
+     * @param string $message
+     * @param null $store
      */
-    protected function getFailedIds()
+    protected function logMessage($message, $store = null)
     {
-        return array_map(function ($failed) {
-            return (int) $failed->getId();
-        }, $this->failed);
+        if ($store) {
+            Mage::helper('payone_core')->logCronjobMessage($message, $store);
+        }
+        else {
+            Mage::helper('payone_core')->logCronjobMessage($message);
+        }
     }
 
     /**
-     * @return int[] The TxIDs of all failed transaction status.
+     * Loads and returns next pending TX status or returns null
+     * if no pending TX status exists.
+     *
+     * @return Payone_Core_Model_Domain_Protocol_TransactionStatus|null
      */
-    protected function getFailedTxIds()
+    protected function loadNextTxStatus()
     {
-        return array_map(function ($failed) {
-            /** @var Payone_Core_Model_Domain_Protocol_TransactionStatus $failed */
-            return (int) $failed->getTxid();
-        }, $this->failed);
+        $collection = $this->getFactory()->getModelTransactionStatus()->getCollection();
+
+        // Get only pending TX status which were not already processed or excluded.
+        $collection->addFieldToFilter('processing_status', Payone_Core_Model_Domain_Protocol_TransactionStatus::STATUS_PENDING);
+
+        if (!empty($this->processedTxStatus)) {
+            $collection->addFieldToFilter('id', array('nin' => array_keys($this->processedTxStatus)));
+        }
+
+        if (!empty($this->excludedTxIds)) {
+            $collection->addFieldToFilter('txid', array('nin' => $this->excludedTxIds));
+        }
+
+        // Make sure we get the most recent TX status to make sure newer status are processed
+        // before any old dangling status.
+        $collection->setOrder('txtime', Payone_Core_Model_Domain_Resource_Protocol_TransactionStatus_Collection::SORT_ORDER_DESC);
+
+        // Limit result to one entity
+        $collection->getSelect()->limit(1);
+
+        $this->logMessage(sprintf('Fetching next TX status -- %s', $collection->getSelectSql(true)));
+
+        $collection->load();
+
+        /** @var Payone_Core_Model_Domain_Protocol_TransactionStatus $txStatus */
+        foreach ($collection as $txStatus) {
+            return $txStatus;
```

Comment:
This looks like a weird construct to me - maybe something like `$collection->getFirstItem()` would be asier?